src/darwin/Framework/CHIP/MTROTAProviderDelegateBridge.mm

```diff
@@ -608,81 +608,82 @@ bool GetPeerNodeInfo(CommandHandler * commandHandler, const ConcreteCommandPath
+                          [[data description] cStringUsingEncoding:NSUTF8StringEncoding]);
 
-                    response.imageURI.SetValue(uri);
-                    handler->AddResponse(cachedCommandPath, response);
-                    handle.Release();
-                    return;
-                }
-                if (!hasUpdate) {
-                    // Send whatever error response our delegate decided on.
-                    handler->AddResponse(cachedCommandPath, response);
-                } else {
-                    // We must have isBDXProtocolSupported false.  Send the corresponding error status.
-                    Commands::QueryImageResponse::Type protocolNotSupportedResponse;
-                    protocolNotSupportedResponse.status
-                        = static_cast<OTAQueryStatus>(MTROTASoftwareUpdateProviderOTAQueryStatusDownloadProtocolNotSupported);
-                    handler->AddResponse(cachedCommandPath, protocolNotSupportedResponse);
-                }
-                handle.Release();
-                gOtaSender.ResetState();
-            }
+                      Commands::QueryImageResponse::Type response;
+
+                      // If BDX protocol is not supported, send a response with status - Protocol Not supported
+                      if (![commandParams.protocolsSupported
+                              containsObject:@(MTROtaSoftwareUpdateProviderOTADownloadProtocolBDXSynchronous)]) {
+                          response.status
+                              = static_cast<OTAQueryStatus>(MTROTASoftwareUpdateProviderOTAQueryStatusDownloadProtocolNotSupported);
+                      } else if ([data.status isEqual:@(MTROtaSoftwareUpdateProviderOTAQueryStatusUpdateAvailable)]) {
+                          ConvertFromQueryImageResponseParms(data, response);
+
+                          // If there is an update available, return a response with status - Available and image uri set in the
+                          // response
+                          auto fabricIndex = handler->GetSubjectDescriptor().fabricIndex;
+                          auto nodeId = handler->GetSubjectDescriptor().subject;
+                          CHIP_ERROR err = gOtaSender.PrepareForTransfer(fabricIndex, nodeId);
+                          if (CHIP_NO_ERROR != err) {
+                              LogErrorOnFailure(err);
+
+                              // Handle busy error separately as we have a query image response status that maps to busy
+                              if (err == CHIP_ERROR_BUSY) {
+                                  response.status = static_cast<OTAQueryStatus>(MTROTASoftwareUpdateProviderOTAQueryStatusBusy);
+                                  if (!response.delayedActionTime.HasValue()) {
+                                      response.delayedActionTime.SetValue(kDelayedActionTimeSeconds);
+                                  }
```

Comment:
Shouldn't this.... send the response?  We _really_ need tests for this code...  Can we please try to stand up some automated tests before making these changes?

Reply:
I think  the below should send the response. I wrote the code to build the response object in all the if/elseif/else case and then the lines at 674 will send the response

handler->AddResponse(cachedCommandPath, response); 
                      handle.Release();

Reply:
Yes, I see that, but that leads to bad behavior like "were going to start constructing BDX URIs even though we are planning to respond Busy", no?

Reply:
yes thats true. and not what I intended. i wanted to do something if (err) {build response } else {build response}and i completely missed the else part. idea was to create the response for each case and have the handler code in common. But nvm i wrote the code inline and uploading and lets see what common code can be moved to apis if it makes sense